java/ql/lib/semmle/code/java/JDK.qll

```diff
@@ -197,6 +203,48 @@ class TypeFile extends Class {
   TypeFile() { this.hasQualifiedName("java.io", "File") }
 }
 
+/**
+ * A taint step from an update of the `bytes[]` parameter in an override of the `InputStream.read` method
+ * to a class instance expression of the type extending `InputStream`.
+ *
+ * This models how a subtype of `InputStream` could be tainted by the definition of its methods, which will
+ * normally only happen in anonymous classes.
+ */
+private class InputStreamWrapperAnonymousStep extends AdditionalTaintStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(Method m, AnonymousClass wrapper |
```

Comment:
Consider using `NestedClass` instead of `AnonymousClass` since then we'll also cover local class definitions (they can also capture variables). Then to get to the `ClassInstanceExpr`, you can replace `n2.asExpr() = wrapper.getClassInstanceExpr()` with `n2.asExpr().(ClassInstanceExpr).getConstructedType().getASourceSupertype*().getSourceDeclaration() = wrapper`.

Reply:
Sounds good! Although isn't `getASourceDeclaration()` redundant after `getASourceSupertype*()`?

Reply:
No, not if we consider the reflexive case, i.e. we take zero `getASourceSupertype` steps.

Reply:
Oh, good point, thanks!